### ISSUE TITLE
allow-underscore-test for CamelCaseMethod gives unexpected results

### ISSUE THREAD

**Issue by SanderSander** [phpmd/phpmd#851]
- PHPMD version: 2.9.1
- PHP Version: 7.4
- Installation type: composer

# Current Behavior
Currently when the property `allow-underscore-test` is set to true for the `CamelCaseMethod` rule, method names like `test_that_something_is_ok` are not accepted while documentation says "Is it allowed to have underscores in test method names." 

Changing the regex in the rule to `/^test[a-zA-Z0-9]*([_][a-z][a-zA-Z0-9]*)*?$/` should fix the problem. 

# Expected behavior
Method like `test_that_something_is_ok` should be allowed as described in documentation

# Additional 
I could make a PR myself, but in the Unit tests the following is documented:

```
    /**
     * Tests that the rule does apply for a test method name
     * with multiple underscores even when one is allowed.
     *
     * @return void
     */
```

So I'm not sure what the expected behavior would be?

**Comment by kylekatarnls** [comment]
Hello,

`test_that_something_is_ok` matches the RegExp and does not produce violation when `allow-underscore-test` is true.

Please provide a minimal way to reproduce your error.

**Comment by kylekatarnls** [comment]
The PHPDoc code you mention is about forbidding multiple underscores, but `test_that_something_is_ok` does not contain multiple underscores, so it falls to `testRuleDoesNotApplyForTestMethodWithUnderscoreWhenAllowed` which is a o-violation test.

**Comment by SanderSander** [comment]
> The PHPDoc code you mention is about forbidding multiple underscores, but `test_that_something_is_ok` does not contain multiple underscores, so it falls to `testRuleDoesNotApplyForTestMethodWithUnderscoreWhenAllowed` which is a o-violation test.

I misinterpreted the documentation in the unit test I think.... Do you mean a method name like `test_hello__oops`

I did some more research but it really isn't working ;) You can test this yourself by changing the file `src/test/resources/files/Rule/Controversial/CamelCaseMethodName/testRuleDoesNotApplyForTestMethodWithUnderscoreWhenAllowed.phptestRuleDoesNotApplyForTestMethodWithUnderscoreWhenAllowed` and change the function name there. 

The test will fail if you use the method name `test_that_something_is_ok`. 

My `ruleset.xml` 

```
<?xml version="1.0"?>
<ruleset name="PHPMD rule set"
         xmlns="http://pmd.sf.net/ruleset/1.0.0"
         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
         xsi:schemaLocation="http://pmd.sf.net/ruleset/1.0.0 http://pmd.sf.net/ruleset_xml_schema.xsd"
         xsi:noNamespaceSchemaLocation="http://pmd.sf.net/ruleset_xml_schema.xsd">

    <rule ref="rulesets/controversial.xml">
        <exclude name="CamelCaseMethodName" />
    </rule>
    <rule ref="rulesets/controversial.xml/CamelCaseMethodName">
        <properties>
            <property name="allow-underscore-test" value="true" />
        </properties>
    </rule>
</ruleset>
```

Furthermore in the test `testRuleAppliesToTestMethodWithTwoUnderscoresEvenWhenOneIsAllowed` the method with name `testGivenSomeValue_expectSome_niceResult` is tested which I expect to be a valid method name and not an invalid method name because like you said it doesn't contain multiple underscores.

**Comment by sebestenyb** [comment]
I think the [regex](https://github.com/phpmd/phpmd/blob/master/src/main/php/PHPMD/Rule/Controversial/CamelCaseMethodName.php#L83) at is still not correct:

```
> preg_match('/^test[a-zA-Z0-9]*(_[a-z][a-zA-Z0-9]*)*$/', 'test_it_throws_if_more_5') 
= 0

> preg_match('/^test[a-zA-Z0-9]*(_[a-z][a-zA-Z0-9]*)*$/', 'test_it_throws_if_more_f5')
= 1
```

Not sure about the role of that extra `[a-z]` block, without it it seems working correctly:

```
> preg_match('/^test[a-zA-Z0-9]*(_[a-zA-Z0-9]*)*$/', 'test_it_throws_if_more_5')      
= 1
```

**Comment by kylekatarnls** [comment]
I guess this block is to disallow uppercase after underscore but disallowing number is probably not intended. [a-z0-9] instead of [a-z] would be fine IMO.
